{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Jun 22 14:09:48 PDT 2022\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DESeq2 - Differential gene expression analysis based on the negative binomial distribution\n",
    "Activate environment:  \n",
    "conda activate diffexpr\n",
    "\n",
    "Cite:  \n",
    "https://github.com/wckdouglas/diffexpr  \n",
    "https://bioconductor.org/packages/release/bioc/html/DESeq2.html  \n",
    "\n",
    "DESeq2 manual:  \n",
    "http://www.bioconductor.org/packages/release/bioc/manuals/DESeq2/man/DESeq2.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%config InlineBackend.figure_format = 'retina'\n",
    "%load_ext blackcellmagic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import anndata\n",
    "# import scvi\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import gget\n",
    "\n",
    "from diffexpr.py_deseq import py_DESeq2\n",
    "\n",
    "from upsetplot import UpSet\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib import cm\n",
    "\n",
    "import glob\n",
    "\n",
    "sc.set_figure_params(figsize=(6, 6), frameon=False)\n",
    "sc.settings.n_jobs=2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set random seed\n",
    "np.random.seed(926)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def nd(arr):\n",
    "    \"\"\"\n",
    "    Function to transform numpy matrix to nd array.\n",
    "    \"\"\"\n",
    "    return np.asarray(arr).reshape(-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 35804 × 22151\n",
       "    obs: 'species', 'batch', 'n_counts_processed', 'batch_index', 'n_counts_raw', 'leiden', 'celltype', 'connectivity'\n",
       "    var: 'gene_name', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'batch_colors', 'celltype_colors', 'hvg', 'leiden', 'leiden_colors', 'neighbors', 'pca', 'rank_genes_groups', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = anndata.read_h5ad(\"../../finchseq_data/all_celltype.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define masks to separate control and experiment datasets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "control_mask = np.logical_or(adata.obs[\"batch\"]==\"control1\", adata.obs[\"batch\"]==\"control2\")\n",
    "experiment_mask = np.logical_or(adata.obs[\"batch\"]==\"experiment1\", adata.obs[\"batch\"]==\"experiment2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add new obs column to separate between control and experiment in general, without separating between batches:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>batch</th>\n",
       "      <th>n_counts_processed</th>\n",
       "      <th>batch_index</th>\n",
       "      <th>n_counts_raw</th>\n",
       "      <th>leiden</th>\n",
       "      <th>celltype</th>\n",
       "      <th>connectivity</th>\n",
       "      <th>batch_g</th>\n",
       "      <th>celltype_g</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCACTGGA-0</th>\n",
       "      <td>zebrafinch</td>\n",
       "      <td>control1</td>\n",
       "      <td>483.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>483.000000</td>\n",
       "      <td>13</td>\n",
       "      <td>mural cells 2</td>\n",
       "      <td>0</td>\n",
       "      <td>control</td>\n",
       "      <td>mural cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGTCAAG-0</th>\n",
       "      <td>zebrafinch</td>\n",
       "      <td>control1</td>\n",
       "      <td>946.000061</td>\n",
       "      <td>0</td>\n",
       "      <td>946.000061</td>\n",
       "      <td>3</td>\n",
       "      <td>microglia 1</td>\n",
       "      <td>1</td>\n",
       "      <td>control</td>\n",
       "      <td>microglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTCACAG-0</th>\n",
       "      <td>zebrafinch</td>\n",
       "      <td>control1</td>\n",
       "      <td>1068.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1068.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>GABAergic neurons 1</td>\n",
       "      <td>2</td>\n",
       "      <td>control</td>\n",
       "      <td>GABAergic neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTCATTGC-0</th>\n",
       "      <td>zebrafinch</td>\n",
       "      <td>control1</td>\n",
       "      <td>1407.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1407.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>astrocytes 2</td>\n",
       "      <td>3</td>\n",
       "      <td>control</td>\n",
       "      <td>astrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTGCTACT-0</th>\n",
       "      <td>zebrafinch</td>\n",
       "      <td>control1</td>\n",
       "      <td>1060.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1060.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>GABAergic neurons 1</td>\n",
       "      <td>2</td>\n",
       "      <td>control</td>\n",
       "      <td>GABAergic neurons</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       species     batch  n_counts_processed batch_index  \\\n",
       "barcode                                                                    \n",
       "AAACCCAAGCACTGGA-0  zebrafinch  control1          483.000000           0   \n",
       "AAACCCAAGCGTCAAG-0  zebrafinch  control1          946.000061           0   \n",
       "AAACCCAAGGTCACAG-0  zebrafinch  control1         1068.000000           0   \n",
       "AAACCCAAGTCATTGC-0  zebrafinch  control1         1407.000000           0   \n",
       "AAACCCAAGTGCTACT-0  zebrafinch  control1         1060.000000           0   \n",
       "\n",
       "                    n_counts_raw leiden             celltype connectivity  \\\n",
       "barcode                                                                     \n",
       "AAACCCAAGCACTGGA-0    483.000000     13        mural cells 2            0   \n",
       "AAACCCAAGCGTCAAG-0    946.000061      3          microglia 1            1   \n",
       "AAACCCAAGGTCACAG-0   1068.000000      1  GABAergic neurons 1            2   \n",
       "AAACCCAAGTCATTGC-0   1407.000000     10         astrocytes 2            3   \n",
       "AAACCCAAGTGCTACT-0   1060.000000      1  GABAergic neurons 1            2   \n",
       "\n",
       "                    batch_g          celltype_g  \n",
       "barcode                                          \n",
       "AAACCCAAGCACTGGA-0  control        mural cells   \n",
       "AAACCCAAGCGTCAAG-0  control          microglia   \n",
       "AAACCCAAGGTCACAG-0  control  GABAergic neurons   \n",
       "AAACCCAAGTCATTGC-0  control         astrocytes   \n",
       "AAACCCAAGTGCTACT-0  control  GABAergic neurons   "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"batch_g\"] = \"\"\n",
    "\n",
    "adata.obs.loc[control_mask, \"batch_g\"] = \"control\"\n",
    "adata.obs.loc[experiment_mask, \"batch_g\"] = \"experiment\"\n",
    "\n",
    "# Create columns containing general celltype assignment - ignoring cluster separation\n",
    "adata.obs['celltype_g'] = adata.obs['celltype'].str.replace('\\d+', '')\n",
    "\n",
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DESeq2 on mean gene expression per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:27 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:27 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:27 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:28 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:28 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:37:28 PM PDT WARNING R[write to console]: Error in lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  : \n",
      "  newsplit: out of vertex space\n",
      "\n",
      "Sat 18 Jun 2022 03:37:28 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:37:28 PM PDT WARNING R[write to console]: There were 25 warnings (use warnings() to see them)\n",
      "Sat 18 Jun 2022 03:37:28 PM PDT WARNING R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mural cells 2 errored out with error: Error in lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  : \n",
      "  newsplit: out of vertex space\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:29 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:29 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:29 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:30 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:30 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:37:30 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:31 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:37:32 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:37:34 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:34 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:34 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: Warning messages:\n",
      "\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: 1: \n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: 2: \n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: In lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  :\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]:  Estimated rdf < 1.0; not estimating variance\n",
      "\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: 3: \n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:35 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GABAergic neurons 1 errored out with error: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:36 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:36 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:36 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:37 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:37 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:37:37 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:39 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:37:40 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:37:41 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:41 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:41 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: Warning messages:\n",
      "\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: 1: \n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: 2: \n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: In lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  :\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]:  Estimated rdf < 1.0; not estimating variance\n",
      "\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: 3: \n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:42 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "astrocytes 1 errored out with error: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:43 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:43 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:44 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:44 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:45 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:46 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:37:47 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:37:48 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:48 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:49 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:49 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:49 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:37:49 PM PDT WARNING R[write to console]: Error in lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  : \n",
      "  newsplit: out of vertex space\n",
      "\n",
      "Sat 18 Jun 2022 03:37:49 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:37:49 PM PDT WARNING R[write to console]: There were 19 warnings (use warnings() to see them)\n",
      "Sat 18 Jun 2022 03:37:49 PM PDT WARNING R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mural / vascular endothelial cells 1 errored out with error: Error in lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  : \n",
      "  newsplit: out of vertex space\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:50 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:51 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:51 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:52 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:52 PM PDT WARNING R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "Sat 18 Jun 2022 03:37:52 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:37:52 PM PDT WARNING R[write to console]: Warning message:\n",
      "\n",
      "Sat 18 Jun 2022 03:37:52 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:37:52 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:52 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "glutamatergic neurons 3 errored out with error: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:52 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:53 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:53 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:54 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:54 PM PDT WARNING R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "Sat 18 Jun 2022 03:37:54 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:37:54 PM PDT WARNING R[write to console]: Warning message:\n",
      "\n",
      "Sat 18 Jun 2022 03:37:54 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:37:54 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:54 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "glutamatergic neurons 1 errored out with error: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:55 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:55 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:55 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:56 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:56 PM PDT WARNING R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "Sat 18 Jun 2022 03:37:56 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:37:56 PM PDT WARNING R[write to console]: Warning message:\n",
      "\n",
      "Sat 18 Jun 2022 03:37:56 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:37:56 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:56 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "migrating neuroblasts errored out with error: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:57 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:57 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:57 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:37:58 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:37:58 PM PDT WARNING R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "Sat 18 Jun 2022 03:37:58 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:37:58 PM PDT WARNING R[write to console]: Warning message:\n",
      "\n",
      "Sat 18 Jun 2022 03:37:58 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:37:58 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:37:58 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GABAergic neurons 2 errored out with error: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:37:59 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:37:59 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:37:59 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:00 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:00 PM PDT WARNING R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "Sat 18 Jun 2022 03:38:00 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:38:00 PM PDT WARNING R[write to console]: Warning message:\n",
      "\n",
      "Sat 18 Jun 2022 03:38:00 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:38:00 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:38:00 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "glutamatergic neurons 2 errored out with error: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:38:01 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:01 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:01 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:02 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:02 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:38:02 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:04 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:38:05 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:38:06 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:06 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:07 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:07 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:07 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:38:08 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:09 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:38:10 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:38:11 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:12 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:12 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:12 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: Warning messages:\n",
      "\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: 1: \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: 2: \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: In lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  :\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]:  Estimated rdf < 1.0; not estimating variance\n",
      "\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: 3: \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: 4: \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: In lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  :\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]:  Estimated rdf < 1.0; not estimating variance\n",
      "\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: 5: \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: \n",
      " \n",
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]:  some variables in design formula are characters, converting to factors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "oligodendrocytes 2 errored out with error: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:38:13 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:14 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:14 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:15 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:15 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:38:15 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:16 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:38:17 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:38:18 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:19 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:19 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:20 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:20 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:38:20 PM PDT WARNING R[write to console]: Error in lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  : \n",
      "  newsplit: out of vertex space\n",
      "\n",
      "Sat 18 Jun 2022 03:38:20 PM PDT WARNING R[write to console]: In addition: \n",
      "Sat 18 Jun 2022 03:38:20 PM PDT WARNING R[write to console]: There were 20 warnings (use warnings() to see them)\n",
      "Sat 18 Jun 2022 03:38:20 PM PDT WARNING R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "oligodendrocytes 1 errored out with error: Error in lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,  : \n",
      "  newsplit: out of vertex space\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:38:21 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:21 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:21 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:22 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:22 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:38:22 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:23 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:38:24 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:38:26 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:26 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:26 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:27 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:27 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:29 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:38:30 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:38:31 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:32 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:32 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:33 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:33 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:38:33 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:34 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:38:35 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:38:36 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:37 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:37 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:38 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:38 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:38:38 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:40 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:38:41 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Sat 18 Jun 2022 03:38:42 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Sat 18 Jun 2022 03:38:43 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Sat 18 Jun 2022 03:38:43 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:43 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Sat 18 Jun 2022 03:38:44 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Sat 18 Jun 2022 03:38:44 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Sat 18 Jun 2022 03:38:45 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Sat 18 Jun 2022 03:38:46 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 33s, sys: 2min 13s, total: 4min 47s\n",
      "Wall time: 1min 21s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "for cluster in adata.obs.celltype.values.unique():\n",
    "    # Create masks to obtain only retain cells from control or experiment, and one cluster\n",
    "    celltype_mask_control1 = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & [adata.obs[\"batch\"]==\"control1\"]\n",
    "    celltype_mask_control2 = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & [adata.obs[\"batch\"]==\"control2\"]\n",
    "    celltype_mask_experiment1 = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & [adata.obs[\"batch\"]==\"experiment1\"]\n",
    "    celltype_mask_experiment2 = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & [adata.obs[\"batch\"]==\"experiment2\"]\n",
    "\n",
    "    # Apply masks to raw data\n",
    "    control_data1 = adata.raw[celltype_mask_control1]\n",
    "    control_data2 = adata.raw[celltype_mask_control2]\n",
    "    experiment_data1 = adata.raw[celltype_mask_experiment1]\n",
    "    experiment_data2 = adata.raw[celltype_mask_experiment2]\n",
    "\n",
    "    # Filter for only highly variable genes\n",
    "    control_data1 = control_data1[:,adata.var['highly_variable']]\n",
    "    control_data2 = control_data2[:,adata.var['highly_variable']]\n",
    "    experiment_data1 = experiment_data1[:,adata.var['highly_variable']]\n",
    "    experiment_data2 = experiment_data2[:,adata.var['highly_variable']]\n",
    "\n",
    "    ## Build count matrix\n",
    "    count_matrix = pd.DataFrame()\n",
    "    count_matrix[\"gene\"] = adata.raw.var[adata.var['highly_variable']==True].index.values\n",
    "\n",
    "    # Get mean gene expression per condition for each gene (rounded to integer)\n",
    "    count_matrix[\"C_1\"] = np.around(control_data1.X.mean(axis=0).A1).astype(int)\n",
    "    count_matrix[\"C_2\"] = np.around(control_data2.X.mean(axis=0).A1).astype(int)\n",
    "    count_matrix[\"E_1\"] = np.around(experiment_data1.X.mean(axis=0).A1).astype(int)\n",
    "    count_matrix[\"E_2\"] = np.around(experiment_data2.X.mean(axis=0).A1).astype(int)\n",
    "\n",
    "    ## Build design matrix\n",
    "    design_matrix = pd.DataFrame()\n",
    "    design_matrix[\"samplename\"] = count_matrix.columns.values[1:]\n",
    "    design_matrix = design_matrix.assign(\n",
    "        sample=lambda d: d.samplename.str.extract(\"([CE])_\", expand=False)\n",
    "    )\n",
    "    design_matrix[\"replicate\"] = design_matrix[\"samplename\"].str.extractall(\"(\\d+)\").values\n",
    "    design_matrix.index = design_matrix.samplename\n",
    "\n",
    "\n",
    "    # Run DESeq2\n",
    "    try:\n",
    "        dds = py_DESeq2(\n",
    "            count_matrix=count_matrix,\n",
    "            design_matrix=design_matrix,\n",
    "            design_formula=\"~ replicate + sample\",\n",
    "            gene_column=\"gene\",\n",
    "        )\n",
    "        dds.run_deseq(\n",
    "            test=\"LRT\",                            # Likelihood ratio test on the difference in deviance between a full and reduced model formula (defined by nbinomLRT)\n",
    "            reduced=\"~ replicate\",                 # Reduced model without condition as variable\n",
    "            # fitType=\"parametric\",                  # Either \"parametric\" (default), \"local\", \"mean\", or \"glmGamPoi\" for the type of fitting of dispersions to the mean intensity\n",
    "            sfType=\"poscounts\",                    # Type of size factor estimation\n",
    "            betaPrior=False,                       # Whether or not to put a zero-mean normal prior on the non-intercept coefficients (default: False)\n",
    "            minReplicatesForReplace=float(\"inf\"),  # The minimum number of replicates required in order to use replaceOutliers on a sample. \n",
    "            useT=True,                             # Default is FALSE, where Wald statistics are assumed to follow a standard Normal\n",
    "            minmu=1e-06,                           # Lower bound on the estimated count for fitting gene-wise dispersion (default: 1e-06 for glmGamPoi)\n",
    "\n",
    "        )\n",
    "        dds.get_deseq_result(contrast=[\"sample\", \"C\", \"E\"])\n",
    "        res = dds.deseq_result\n",
    "\n",
    "        # Add column with cluster name\n",
    "        res[\"cluster\"] = cluster\n",
    "        \n",
    "        res.to_csv(f\"deseq2_mean_results/deseq2_mean_{cluster.replace(' ', '-').replace('/-', '')}.csv\")\n",
    "    \n",
    "    except Exception as e:\n",
    "        print(cluster + \" errored out with error: \" + str(e))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get a data frame with the significant p-values for all clusters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "for csv in glob.glob(\"deseq2_mean_results/*.csv\"):\n",
    "    df = pd.read_csv(csv)\n",
    "    if len(df[df[\"padj\"] < 0.05][\"padj\"]) > 0:\n",
    "        if i == 0:\n",
    "            df_sig = df[df[\"padj\"] < 0.05].copy()\n",
    "            i = +1\n",
    "        else:\n",
    "            df_sig = df_sig.append(df[df[\"padj\"] < 0.05], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>gene</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JCHAIN_ENSTGUG00000001661.2</td>\n",
       "      <td>4.75</td>\n",
       "      <td>-18.585862</td>\n",
       "      <td>191.135443</td>\n",
       "      <td>15.587912</td>\n",
       "      <td>7.875649e-05</td>\n",
       "      <td>1.785147e-03</td>\n",
       "      <td>JCHAIN_ENSTGUG00000001661.2</td>\n",
       "      <td>microglia 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>_ENSTGUG00000028479.1</td>\n",
       "      <td>78.50</td>\n",
       "      <td>-6.326571</td>\n",
       "      <td>1.437911</td>\n",
       "      <td>16.014314</td>\n",
       "      <td>6.286539e-05</td>\n",
       "      <td>1.785147e-03</td>\n",
       "      <td>_ENSTGUG00000028479.1</td>\n",
       "      <td>microglia 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>_ENSTGUG00000028531.1</td>\n",
       "      <td>15.00</td>\n",
       "      <td>-20.132631</td>\n",
       "      <td>184.945221</td>\n",
       "      <td>22.794176</td>\n",
       "      <td>1.803113e-06</td>\n",
       "      <td>1.226117e-04</td>\n",
       "      <td>_ENSTGUG00000028531.1</td>\n",
       "      <td>microglia 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>_ENSTGUG00000027913.1</td>\n",
       "      <td>8.00</td>\n",
       "      <td>20.075439</td>\n",
       "      <td>250.101009</td>\n",
       "      <td>37.619242</td>\n",
       "      <td>8.599199e-10</td>\n",
       "      <td>3.207501e-07</td>\n",
       "      <td>_ENSTGUG00000027913.1</td>\n",
       "      <td>oligodendrocytes 3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0  baseMean  log2FoldChange       lfcSE  \\\n",
       "0  JCHAIN_ENSTGUG00000001661.2      4.75      -18.585862  191.135443   \n",
       "1        _ENSTGUG00000028479.1     78.50       -6.326571    1.437911   \n",
       "2        _ENSTGUG00000028531.1     15.00      -20.132631  184.945221   \n",
       "3        _ENSTGUG00000027913.1      8.00       20.075439  250.101009   \n",
       "\n",
       "        stat        pvalue          padj                         gene  \\\n",
       "0  15.587912  7.875649e-05  1.785147e-03  JCHAIN_ENSTGUG00000001661.2   \n",
       "1  16.014314  6.286539e-05  1.785147e-03        _ENSTGUG00000028479.1   \n",
       "2  22.794176  1.803113e-06  1.226117e-04        _ENSTGUG00000028531.1   \n",
       "3  37.619242  8.599199e-10  3.207501e-07        _ENSTGUG00000027913.1   \n",
       "\n",
       "              cluster  \n",
       "0         microglia 2  \n",
       "1         microglia 2  \n",
       "2         microglia 2  \n",
       "3  oligodendrocytes 3  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sat 18 Jun 2022 03:39:41 PM PDT INFO We noticed that you may have passed a version number with your Ensembl ID.\n",
      "Please note that gget info will always return information linked to the latest Ensembl ID version (see 'ensembl_id').\n",
      "Sat 18 Jun 2022 03:39:48 PM PDT WARNING No reviewed UniProt results were found for Ensembl ID ENSTGUG00000001661. Returning all unreviewed results.\n",
      "Sat 18 Jun 2022 03:39:52 PM PDT WARNING No reviewed UniProt results were found for Ensembl ID ENSTGUG00000028479. Returning all unreviewed results.\n",
      "Sat 18 Jun 2022 03:39:55 PM PDT WARNING No reviewed UniProt results were found for Ensembl ID ENSTGUG00000028531. Returning all unreviewed results.\n",
      "Sat 18 Jun 2022 03:39:58 PM PDT WARNING No reviewed UniProt results were found for Ensembl ID ENSTGUG00000027913. Returning all unreviewed results.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>primary_gene_name</th>\n",
       "      <th>ensembl_gene_name</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>parent_gene</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>ensembl_description</th>\n",
       "      <th>uniprot_description</th>\n",
       "      <th>ncbi_description</th>\n",
       "      <th>object_type</th>\n",
       "      <th>biotype</th>\n",
       "      <th>canonical_transcript</th>\n",
       "      <th>seq_region_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>all_transcripts</th>\n",
       "      <th>transcript_biotypes</th>\n",
       "      <th>transcript_names</th>\n",
       "      <th>transcript_strands</th>\n",
       "      <th>transcript_starts</th>\n",
       "      <th>transcript_ends</th>\n",
       "      <th>all_exons</th>\n",
       "      <th>exon_starts</th>\n",
       "      <th>exon_ends</th>\n",
       "      <th>all_translations</th>\n",
       "      <th>translation_starts</th>\n",
       "      <th>translation_ends</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSTGUG00000001661</th>\n",
       "      <td>ENSTGUG00000001661.2</td>\n",
       "      <td>[A0A674GW10, A0A674GX51, H0YTU6]</td>\n",
       "      <td>100217545</td>\n",
       "      <td>taeniopygia_guttata</td>\n",
       "      <td>bTaeGut1_v1.p</td>\n",
       "      <td>JCHAIN</td>\n",
       "      <td>JCHAIN</td>\n",
       "      <td>[IGJ, JCHAIN]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>Taeniopygia guttata joining<br>chain of multimeric IgA and<br>IgM (JCHAIN), mRNA.<br>[Source:RefSeq<br>mRNA;Acc:NM_001309502]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSTGUT00000027687.1</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>67313744</td>\n",
       "      <td>67318861</td>\n",
       "      <td>[ENSTGUT00000001728.2, ENSTGUT00000039414.1, ENSTGUT00000027687.1]</td>\n",
       "      <td>[protein_coding, protein_coding, protein_coding]</td>\n",
       "      <td>[JCHAIN-201, JCHAIN-202, JCHAIN-203]</td>\n",
       "      <td>[-1, -1, -1]</td>\n",
       "      <td>[67313744, 67313745, 67314471]</td>\n",
       "      <td>[67318861, 67318767, 67316295]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSTGUG00000028479</th>\n",
       "      <td>ENSTGUG00000028479.1</td>\n",
       "      <td>[A0A674GB31, A0A674GBM2, A0A674GCD7, A0A674GHK5, A0A674GNN0, A0A674H7B8, A0A674HBQ1, A0A674HFI2, H0ZF58]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>taeniopygia_guttata</td>\n",
       "      <td>bTaeGut1_v1.p</td>\n",
       "      <td>[nan, nan, nan, nan, nan, nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[nan, nan, nan, nan, nan, nan, nan, nan, nan]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Ig-like domain-containing protein, Uncharacterized protein]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSTGUT00000029820.1</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>5648389</td>\n",
       "      <td>5671784</td>\n",
       "      <td>[ENSTGUT00000029820.1, ENSTGUT00000028177.1, ENSTGUT00000031159.1, ENSTGUT00000022357.1, ENSTGUT00000026496.1, ENSTGUT00000020849.1, ENSTGUT00000035004.1, ENSTGUT00000023157.1, ENSTGUT00000009319.2]</td>\n",
       "      <td>[protein_coding, protein_coding, protein_coding, protein_coding, protein_coding, protein_coding, protein_coding, protein_coding, protein_coding]</td>\n",
       "      <td>[nan, nan, nan, nan, nan, nan, nan, nan, nan]</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1]</td>\n",
       "      <td>[5648389, 5648389, 5652286, 5652286, 5652286, 5656220, 5656675, 5657506, 5666662]</td>\n",
       "      <td>[5671784, 5671784, 5671784, 5671784, 5671784, 5671784, 5671784, 5663165, 5671784]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSTGUG00000028531</th>\n",
       "      <td>ENSTGUG00000028531.1</td>\n",
       "      <td>A0A674H0A8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>taeniopygia_guttata</td>\n",
       "      <td>bTaeGut1_v1.p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ig-like domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSTGUT00000025236.1</td>\n",
       "      <td>RRCB01000107.1</td>\n",
       "      <td>1</td>\n",
       "      <td>976755</td>\n",
       "      <td>1081424</td>\n",
       "      <td>[ENSTGUT00000025236.1]</td>\n",
       "      <td>[protein_coding]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[976755]</td>\n",
       "      <td>[1081424]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSTGUG00000027913</th>\n",
       "      <td>ENSTGUG00000027913.1</td>\n",
       "      <td>A0A674HVK1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>taeniopygia_guttata</td>\n",
       "      <td>bTaeGut1_v1.p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aldo_ket_red domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSTGUT00000024158.1</td>\n",
       "      <td>RRCB01000090.1</td>\n",
       "      <td>1</td>\n",
       "      <td>2120170</td>\n",
       "      <td>2139914</td>\n",
       "      <td>[ENSTGUT00000024158.1]</td>\n",
       "      <td>[protein_coding]</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[2120170]</td>\n",
       "      <td>[2139914]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Clean up gene name so only Ensembl ID is retained\n",
    "de_genes = []\n",
    "for gene in df_sig[\"gene\"]:\n",
    "    de_genes.append(gene.split(\"_\")[1])\n",
    "    \n",
    "# Pass Ensembl ID to gget info\n",
    "df = gget.info(de_genes, wrap_text=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the means with DESeq2 did not work very well, I only got 4 DE genes in total. I think part of the problem is that DESeq2 expects raw data and therefore integers, so I had to round the mean values to the closest integer (often 0). Some celltypes did not run at all because all gene-wise dispersion estimates were within 2 orders of magnitude."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DESeq2 on individual cells as replicates - build loop components"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate DESeq2 count matrix from raw counts (DESeq2 only allows integer values). Each cell will be treated as a replicate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create masks to obtain only retain cells from control or experiment, and one cluster\n",
    "celltype_mask_control = np.char.startswith(nd(adata.obs.celltype.values).astype(str), \"microglia 1\") & np.logical_or(adata.obs[\"batch\"]==\"control1\", adata.obs[\"batch\"]==\"control2\")\n",
    "celltype_mask_experiment = np.char.startswith(nd(adata.obs.celltype.values).astype(str), \"microglia 1\") & np.logical_or(adata.obs[\"batch\"]==\"experiment1\", adata.obs[\"batch\"]==\"experiment2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply masks to raw data\n",
    "control_data = adata.raw[celltype_mask_control]\n",
    "experiment_data = adata.raw[celltype_mask_experiment]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for only highly variable genes\n",
    "control_data = control_data[:,adata.var['highly_variable']]\n",
    "experiment_data = experiment_data[:,adata.var['highly_variable']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>C_1</th>\n",
       "      <th>C_2</th>\n",
       "      <th>C_3</th>\n",
       "      <th>C_4</th>\n",
       "      <th>C_5</th>\n",
       "      <th>C_6</th>\n",
       "      <th>C_7</th>\n",
       "      <th>C_8</th>\n",
       "      <th>C_9</th>\n",
       "      <th>...</th>\n",
       "      <th>E_91</th>\n",
       "      <th>E_92</th>\n",
       "      <th>E_93</th>\n",
       "      <th>E_94</th>\n",
       "      <th>E_95</th>\n",
       "      <th>E_96</th>\n",
       "      <th>E_97</th>\n",
       "      <th>E_98</th>\n",
       "      <th>E_99</th>\n",
       "      <th>E_100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WRE_WRE</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TMEM74_ENSTGUG00000022792.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TMEM170B_ENSTGUG00000025839.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>_ENSTGUG00000020832.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>_ENSTGUG00000020221.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4479</th>\n",
       "      <td>_ENSTGUG00000021313.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4480</th>\n",
       "      <td>KIF5A_ENSTGUG00000025926.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4481</th>\n",
       "      <td>B4GALNT1_ENSTGUG00000028363.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4482</th>\n",
       "      <td>AGAP2_ENSTGUG00000019988.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4483</th>\n",
       "      <td>_ENSTGUG00000022543.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4484 rows × 201 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
       "0                           WRE_WRE    0    0    0    0    0    0    0    0   \n",
       "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
       "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
       "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
       "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    0   \n",
       "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "4479          _ENSTGUG00000021313.1    0    0    0    0    1    0    0    0   \n",
       "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
       "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
       "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    0    0    0    0   \n",
       "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
       "\n",
       "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
       "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "1       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "2       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "3       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "4       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
       "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "4482    0  ...     2     0     0     0     0     0     0     0     0      0  \n",
       "4483    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
       "\n",
       "[4484 rows x 201 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsampling_runs = 10\n",
    "subsample_size = 100\n",
    "\n",
    "count_matrix = pd.DataFrame()\n",
    "count_matrix[\"gene\"] = adata.raw.var[adata.var['highly_variable']==True].index.values\n",
    "\n",
    "## Get expression of all genes for each cell from both control datasets\n",
    "count_matrix_control = pd.DataFrame(control_data.X.todense().T.astype(int))\n",
    "# Randombly subsample contol data\n",
    "count_matrix_control = count_matrix_control.sample(n=subsample_size, axis=\"columns\")\n",
    "# Relabel columns \n",
    "count_matrix_control.columns = np.arange(100)+1\n",
    "# Add prefix to mark these cells as control\n",
    "count_matrix_control = count_matrix_control.add_prefix(\"C_\")\n",
    "\n",
    "## Get expression of all genes for each cell from both experiment datasets\n",
    "count_matrix_experiment = pd.DataFrame(experiment_data.X.todense().T.astype(int))\n",
    "# Randombly subsample experiment data\n",
    "count_matrix_experiment = count_matrix_experiment.sample(n=subsample_size, axis=\"columns\")\n",
    "# Relabel columns \n",
    "count_matrix_experiment.columns = np.arange(100)+1\n",
    "# Add prefix to mark these cells as experiment\n",
    "count_matrix_experiment = count_matrix_experiment.add_prefix(\"E_\")\n",
    "\n",
    "# Concatenate experiment and control data\n",
    "count_matrix = pd.concat(\n",
    "    [count_matrix, count_matrix_control, count_matrix_experiment], axis=1\n",
    ")\n",
    "count_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate DESeq2 design matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samplename</th>\n",
       "      <th>sample</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samplename</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C_1</th>\n",
       "      <td>C_1</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C_2</th>\n",
       "      <td>C_2</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C_3</th>\n",
       "      <td>C_3</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C_4</th>\n",
       "      <td>C_4</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C_5</th>\n",
       "      <td>C_5</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E_96</th>\n",
       "      <td>E_96</td>\n",
       "      <td>E</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E_97</th>\n",
       "      <td>E_97</td>\n",
       "      <td>E</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E_98</th>\n",
       "      <td>E_98</td>\n",
       "      <td>E</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E_99</th>\n",
       "      <td>E_99</td>\n",
       "      <td>E</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E_100</th>\n",
       "      <td>E_100</td>\n",
       "      <td>E</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           samplename sample replicate\n",
       "samplename                            \n",
       "C_1               C_1      C         1\n",
       "C_2               C_2      C         2\n",
       "C_3               C_3      C         3\n",
       "C_4               C_4      C         4\n",
       "C_5               C_5      C         5\n",
       "...               ...    ...       ...\n",
       "E_96             E_96      E        96\n",
       "E_97             E_97      E        97\n",
       "E_98             E_98      E        98\n",
       "E_99             E_99      E        99\n",
       "E_100           E_100      E       100\n",
       "\n",
       "[200 rows x 3 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "design_matrix = pd.DataFrame()\n",
    "design_matrix[\"samplename\"] = count_matrix.columns.values[1:]\n",
    "design_matrix = design_matrix.assign(\n",
    "    sample=lambda d: d.samplename.str.extract(\"([CE])_\", expand=False)\n",
    ")\n",
    "design_matrix[\"replicate\"] = design_matrix[\"samplename\"].str.extractall(\"(\\d+)\").values\n",
    "design_matrix.index = design_matrix.samplename\n",
    "design_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run DESeq2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Thu 16 Jun 2022 01:48:37 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 16 Jun 2022 01:48:37 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 16 Jun 2022 01:48:37 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 16 Jun 2022 01:51:43 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 16 Jun 2022 01:51:43 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Thu 16 Jun 2022 01:51:43 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 16 Jun 2022 02:00:43 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 16 Jun 2022 02:09:48 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 8h 23min 4s, sys: 8h 25min 37s, total: 16h 48min 41s\n",
      "Wall time: 21min 16s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "dds = py_DESeq2(\n",
    "    count_matrix=count_matrix,\n",
    "    design_matrix=design_matrix,\n",
    "    design_formula=\"~ replicate + sample\",\n",
    "    gene_column=\"gene\",\n",
    ")\n",
    "dds.run_deseq(\n",
    "    test=\"LRT\",                            # Likelihood ratio test on the difference in deviance between a full and reduced model formula (defined by nbinomLRT)\n",
    "    reduced=\"~ replicate\",                 # Reduced model without condition as variable\n",
    "    # fitType=\"parametric\",                # Either \"parametric\", \"local\", \"mean\", or \"glmGamPoi\" for the type of fitting of dispersions to the mean intensity\n",
    "    sfType=\"poscounts\",                    # Type of size factor estimation\n",
    "    betaPrior=False,                       # Whether or not to put a zero-mean normal prior on the non-intercept coefficients (default: False)\n",
    "    minReplicatesForReplace=float(\"inf\"),  # The minimum number of replicates required in order to use replaceOutliers on a sample. \n",
    "    useT=True,                             # Default is FALSE, where Wald statistics are assumed to follow a standard Normal\n",
    "    minmu=1e-06,                           # Lower bound on the estimated count for fitting gene-wise dispersion (default: 1e-06 for glmGamPoi)\n",
    " \n",
    ")\n",
    "dds.get_deseq_result(contrast=[\"sample\", \"C\", \"E\"])\n",
    "res = dds.deseq_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WRE_WRE</th>\n",
       "      <td>0.010030</td>\n",
       "      <td>11.700454</td>\n",
       "      <td>59.028487</td>\n",
       "      <td>0.096264</td>\n",
       "      <td>0.756360</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WRE_WRE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM74_ENSTGUG00000022792.1</th>\n",
       "      <td>0.005026</td>\n",
       "      <td>10.716314</td>\n",
       "      <td>62.629328</td>\n",
       "      <td>0.047275</td>\n",
       "      <td>0.827874</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TMEM74_ENSTGUG00000022792.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM170B_ENSTGUG00000025839.1</th>\n",
       "      <td>0.025015</td>\n",
       "      <td>-5.751915</td>\n",
       "      <td>13.670826</td>\n",
       "      <td>0.027882</td>\n",
       "      <td>0.867387</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TMEM170B_ENSTGUG00000025839.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000020832.1</th>\n",
       "      <td>0.010034</td>\n",
       "      <td>11.703627</td>\n",
       "      <td>59.019681</td>\n",
       "      <td>0.096315</td>\n",
       "      <td>0.756298</td>\n",
       "      <td>1.0</td>\n",
       "      <td>_ENSTGUG00000020832.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000020221.1</th>\n",
       "      <td>0.060168</td>\n",
       "      <td>-4.831429</td>\n",
       "      <td>9.907635</td>\n",
       "      <td>0.023088</td>\n",
       "      <td>0.879230</td>\n",
       "      <td>1.0</td>\n",
       "      <td>_ENSTGUG00000020221.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000021313.1</th>\n",
       "      <td>0.015038</td>\n",
       "      <td>12.269741</td>\n",
       "      <td>57.479802</td>\n",
       "      <td>0.145430</td>\n",
       "      <td>0.702941</td>\n",
       "      <td>1.0</td>\n",
       "      <td>_ENSTGUG00000021313.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIF5A_ENSTGUG00000025926.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KIF5A_ENSTGUG00000025926.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B4GALNT1_ENSTGUG00000028363.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B4GALNT1_ENSTGUG00000028363.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP2_ENSTGUG00000019988.1</th>\n",
       "      <td>0.035050</td>\n",
       "      <td>-9.238563</td>\n",
       "      <td>17.473478</td>\n",
       "      <td>0.131692</td>\n",
       "      <td>0.716684</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AGAP2_ENSTGUG00000019988.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000022543.1</th>\n",
       "      <td>0.085121</td>\n",
       "      <td>6.409037</td>\n",
       "      <td>7.842045</td>\n",
       "      <td>0.124557</td>\n",
       "      <td>0.724144</td>\n",
       "      <td>1.0</td>\n",
       "      <td>_ENSTGUG00000022543.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4484 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               baseMean  log2FoldChange      lfcSE      stat  \\\n",
       "WRE_WRE                        0.010030       11.700454  59.028487  0.096264   \n",
       "TMEM74_ENSTGUG00000022792.1    0.005026       10.716314  62.629328  0.047275   \n",
       "TMEM170B_ENSTGUG00000025839.1  0.025015       -5.751915  13.670826  0.027882   \n",
       "_ENSTGUG00000020832.1          0.010034       11.703627  59.019681  0.096315   \n",
       "_ENSTGUG00000020221.1          0.060168       -4.831429   9.907635  0.023088   \n",
       "...                                 ...             ...        ...       ...   \n",
       "_ENSTGUG00000021313.1          0.015038       12.269741  57.479802  0.145430   \n",
       "KIF5A_ENSTGUG00000025926.1     0.000000             NaN        NaN       NaN   \n",
       "B4GALNT1_ENSTGUG00000028363.1  0.000000             NaN        NaN       NaN   \n",
       "AGAP2_ENSTGUG00000019988.1     0.035050       -9.238563  17.473478  0.131692   \n",
       "_ENSTGUG00000022543.1          0.085121        6.409037   7.842045  0.124557   \n",
       "\n",
       "                                 pvalue  padj                           gene  \n",
       "WRE_WRE                        0.756360   1.0                        WRE_WRE  \n",
       "TMEM74_ENSTGUG00000022792.1    0.827874   1.0    TMEM74_ENSTGUG00000022792.1  \n",
       "TMEM170B_ENSTGUG00000025839.1  0.867387   1.0  TMEM170B_ENSTGUG00000025839.1  \n",
       "_ENSTGUG00000020832.1          0.756298   1.0          _ENSTGUG00000020832.1  \n",
       "_ENSTGUG00000020221.1          0.879230   1.0          _ENSTGUG00000020221.1  \n",
       "...                                 ...   ...                            ...  \n",
       "_ENSTGUG00000021313.1          0.702941   1.0          _ENSTGUG00000021313.1  \n",
       "KIF5A_ENSTGUG00000025926.1          NaN   NaN     KIF5A_ENSTGUG00000025926.1  \n",
       "B4GALNT1_ENSTGUG00000028363.1       NaN   NaN  B4GALNT1_ENSTGUG00000028363.1  \n",
       "AGAP2_ENSTGUG00000019988.1     0.716684   1.0     AGAP2_ENSTGUG00000019988.1  \n",
       "_ENSTGUG00000022543.1          0.724144   1.0          _ENSTGUG00000022543.1  \n",
       "\n",
       "[4484 rows x 7 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000004607.2</th>\n",
       "      <td>2.686180</td>\n",
       "      <td>-1.457530</td>\n",
       "      <td>0.315486</td>\n",
       "      <td>15.346645</td>\n",
       "      <td>8.947966e-05</td>\n",
       "      <td>0.048006</td>\n",
       "      <td>_ENSTGUG00000004607.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD74_ENSTGUG00000000882.2</th>\n",
       "      <td>15.337836</td>\n",
       "      <td>-0.807027</td>\n",
       "      <td>0.149539</td>\n",
       "      <td>24.534563</td>\n",
       "      <td>7.298874e-07</td>\n",
       "      <td>0.000470</td>\n",
       "      <td>CD74_ENSTGUG00000000882.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000028479.1</th>\n",
       "      <td>12.372307</td>\n",
       "      <td>-17.922297</td>\n",
       "      <td>14.393935</td>\n",
       "      <td>3427.974502</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>_ENSTGUG00000028479.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000019883.1</th>\n",
       "      <td>11.579027</td>\n",
       "      <td>-1.210667</td>\n",
       "      <td>0.183098</td>\n",
       "      <td>34.744482</td>\n",
       "      <td>3.759428e-09</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>_ENSTGUG00000019883.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000029343.1</th>\n",
       "      <td>9.966743</td>\n",
       "      <td>-17.773859</td>\n",
       "      <td>15.224227</td>\n",
       "      <td>2761.491057</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>_ENSTGUG00000029343.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_ENSTGUG00000023468.1</th>\n",
       "      <td>3.142643</td>\n",
       "      <td>-1.383097</td>\n",
       "      <td>0.227779</td>\n",
       "      <td>31.968212</td>\n",
       "      <td>1.567162e-08</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>_ENSTGUG00000023468.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            baseMean  log2FoldChange      lfcSE         stat  \\\n",
       "_ENSTGUG00000004607.2       2.686180       -1.457530   0.315486    15.346645   \n",
       "CD74_ENSTGUG00000000882.2  15.337836       -0.807027   0.149539    24.534563   \n",
       "_ENSTGUG00000028479.1      12.372307      -17.922297  14.393935  3427.974502   \n",
       "_ENSTGUG00000019883.1      11.579027       -1.210667   0.183098    34.744482   \n",
       "_ENSTGUG00000029343.1       9.966743      -17.773859  15.224227  2761.491057   \n",
       "_ENSTGUG00000023468.1       3.142643       -1.383097   0.227779    31.968212   \n",
       "\n",
       "                                 pvalue      padj                       gene  \n",
       "_ENSTGUG00000004607.2      8.947966e-05  0.048006      _ENSTGUG00000004607.2  \n",
       "CD74_ENSTGUG00000000882.2  7.298874e-07  0.000470  CD74_ENSTGUG00000000882.2  \n",
       "_ENSTGUG00000028479.1      0.000000e+00  0.000000      _ENSTGUG00000028479.1  \n",
       "_ENSTGUG00000019883.1      3.759428e-09  0.000004      _ENSTGUG00000019883.1  \n",
       "_ENSTGUG00000029343.1      0.000000e+00  0.000000      _ENSTGUG00000029343.1  \n",
       "_ENSTGUG00000023468.1      1.567162e-08  0.000013      _ENSTGUG00000023468.1  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show significant genes\n",
    "res[res[\"padj\"]<0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# res.to_csv(\"deseq2_micro_test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show normalized gene counts\n",
    "# dds.normalized_count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show coefficients for GLM:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dds.comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # from the last cell, we see the arrangement of coefficients, \n",
    "# # so that we can now use \"coef\" for lfcShrink\n",
    "# # the comparison we want to focus on is 'sample_E_vs_C', so coef = 195 will be used\n",
    "# lfc_res = dds.lfcShrink(coef=195, method='apeglm')\n",
    "# lfc_res.to_csv(\"deseq2_gaba_subsample_lfcShrink.csv\")\n",
    "# lfc_res.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DESeq2 with individual cells as replicates - Loop over all clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "glutamatergic neurons 1                 4144\n",
       "GABAergic neurons 1                     4084\n",
       "migrating neuroblasts                   3336\n",
       "microglia 1                             3039\n",
       "glutamatergic neurons 2                 2846\n",
       "radial glia 1                           2449\n",
       "mural / vascular endothelial cells 1    2377\n",
       "astrocytes 1                            2127\n",
       "glutamatergic neurons 3                 1990\n",
       "microglia 2                             1924\n",
       "astrocytes 2                            1695\n",
       "oligodendrocytes 1                      1570\n",
       "GABAergic neurons 2                     1508\n",
       "mural cells 2                            821\n",
       "oligodendrocyte precursor cells          474\n",
       "radial glia 2                            407\n",
       "red blood cells                          359\n",
       "glutamatergic neurons 4                  296\n",
       "oligodendrocytes 2                       213\n",
       "oligodendrocytes 3                        61\n",
       "GABAergic neurons 3                       59\n",
       "microglia 3 / radial glia                 25\n",
       "Name: celltype, dtype: int64"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show cluster sizes\n",
    "adata.obs.celltype.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# # Define number of subsampling runs\n",
    "# subsampling_runs = 10\n",
    "\n",
    "# for cluster in adata.obs.celltype.values.unique():   \n",
    "#     # Define subsample size\n",
    "#     cluster_size = len(adata.obs[adata.obs[\"celltype\"]==cluster])\n",
    "#     # If cluster includes > 400 cells, subsample size = 100 cells per condition\n",
    "#     if cluster_size >= 400:\n",
    "#         subsample_size = 100\n",
    "#     # If cluster includes < 400 cells, do not subsample\n",
    "#     elif cluster_size < 400 and cluster_size > 100:\n",
    "#         subsample_size = False\n",
    "#     # Exclude clusters with < 100 cells\n",
    "#     else:\n",
    "#         continue\n",
    "    \n",
    "#     print(f\"Running DESeq2 on celltype cluster {cluster} with subsample size {subsample_size}.\")\n",
    "        \n",
    "#     for i in np.arange(subsampling_runs):\n",
    "#         # Create masks to only retain cells from control or experiment, and one cluster\n",
    "#         celltype_mask_control = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & np.logical_or(adata.obs[\"batch\"]==\"control1\", adata.obs[\"batch\"]==\"control2\")\n",
    "#         celltype_mask_experiment = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & np.logical_or(adata.obs[\"batch\"]==\"experiment1\", adata.obs[\"batch\"]==\"experiment2\")\n",
    "\n",
    "#         # Apply masks to raw data\n",
    "#         control_data = adata.raw[celltype_mask_control]\n",
    "#         experiment_data = adata.raw[celltype_mask_experiment]\n",
    "\n",
    "#         # Filter for only highly variable genes\n",
    "#         control_data = control_data[:,adata.var['highly_variable']]\n",
    "#         experiment_data = experiment_data[:,adata.var['highly_variable']]\n",
    "\n",
    "#         ### Create count matrix\n",
    "#         count_matrix = pd.DataFrame()\n",
    "#         count_matrix[\"gene\"] = adata.raw.var[adata.var['highly_variable']==True].index.values\n",
    "\n",
    "#         ## Create matrix including raw expression of all highly variable genes for each cell from both control animals\n",
    "#         count_matrix_control = pd.DataFrame(control_data.X.todense().T.astype(int))\n",
    "        \n",
    "#         if subsample_size:\n",
    "#             # Randombly subsample contol data\n",
    "#             count_matrix_control = count_matrix_control.sample(n=subsample_size, axis=\"columns\")\n",
    "#             # Relabel columns \n",
    "#             count_matrix_control.columns = np.arange(subsample_size)+1\n",
    "            \n",
    "#         # Add prefix to mark these cells as control\n",
    "#         count_matrix_control = count_matrix_control.add_prefix(\"C_\")\n",
    "\n",
    "#         ## Get raw expression of all highly variable genes for each cell from both experiment animals\n",
    "#         count_matrix_experiment = pd.DataFrame(experiment_data.X.todense().T.astype(int))\n",
    "        \n",
    "#         if subsample_size:\n",
    "#             # Randombly subsample experiment data\n",
    "#             count_matrix_experiment = count_matrix_experiment.sample(n=subsample_size, axis=\"columns\")\n",
    "#             # Relabel columns \n",
    "#             count_matrix_experiment.columns = np.arange(subsample_size)+1\n",
    "            \n",
    "#         # Add prefix to mark these cells as experiment\n",
    "#         count_matrix_experiment = count_matrix_experiment.add_prefix(\"E_\")\n",
    "\n",
    "#         ## Concatenate experiment and control data\n",
    "#         count_matrix = pd.concat(\n",
    "#             [count_matrix, count_matrix_control, count_matrix_experiment], axis=1\n",
    "#         )\n",
    "\n",
    "#         # Show the first count matrix for this cluster\n",
    "#         if i==0:\n",
    "#             print(count_matrix)\n",
    "\n",
    "#         ### Create design matrix\n",
    "#         design_matrix = pd.DataFrame()\n",
    "#         design_matrix[\"samplename\"] = count_matrix.columns.values[1:]\n",
    "#         # Get sample ID and replicate number from count matrix column names\n",
    "#         design_matrix = design_matrix.assign(\n",
    "#             sample=lambda d: d.samplename.str.extract(\"([CE])_\", expand=False)\n",
    "#         )\n",
    "#         design_matrix[\"replicate\"] = design_matrix[\"samplename\"].str.extractall(\"(\\d+)\").values\n",
    "#         # Set sample name as index\n",
    "#         design_matrix.index = design_matrix.samplename\n",
    "\n",
    "#         # Show the first design matrix for this cluster\n",
    "#         if i==0:\n",
    "#             print(design_matrix)\n",
    "\n",
    "#         ### Run DESeq2\n",
    "#         dds = py_DESeq2(\n",
    "#             count_matrix=count_matrix,\n",
    "#             design_matrix=design_matrix,\n",
    "#             design_formula=\"~ replicate + sample\",\n",
    "#             gene_column=\"gene\",\n",
    "#         )\n",
    "#         dds.run_deseq(\n",
    "#             test=\"LRT\",                            # Likelihood ratio test on the difference in deviance between a full and reduced model formula (defined by nbinomLRT)\n",
    "#             reduced=\"~ replicate\",                 # Reduced model without condition as variable\n",
    "#             # fitType=\"parametric\",                # Either \"parametric\", \"local\", \"mean\", or \"glmGamPoi\" for the type of fitting of dispersions to the mean intensity (default: parametric)\n",
    "#             sfType=\"poscounts\",                    # Type of size factor estimation\n",
    "#             betaPrior=False,                       # Whether or not to put a zero-mean normal prior on the non-intercept coefficients (default: False)\n",
    "#             minReplicatesForReplace=float(\"inf\"),  # The minimum number of replicates required in order to use replaceOutliers on a sample. \n",
    "#             useT=True,                             # Default is FALSE, where Wald statistics are assumed to follow a standard Normal\n",
    "#             minmu=1e-06,                           # Lower bound on the estimated count for fitting gene-wise dispersion (default: 1e-06 for glmGamPoi)\n",
    "\n",
    "#         )\n",
    "#         dds.get_deseq_result(contrast=[\"sample\", \"C\", \"E\"])\n",
    "#         res = dds.deseq_result\n",
    "        \n",
    "#         res.to_csv(f\"deseq2_results/deseq2_{cluster.replace(' ', '-').replace('/-', '')}_subsample_{i}.csv\") # Results folder renamed to deseq2_results_v1\n",
    "        \n",
    "#         # Do not run again if cluster was not subsampled\n",
    "#         if subsample_size is False:\n",
    "#             break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Redo with equal sampling from each replicate (deseq2_results_v2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes = [\n",
    " 'red blood cells',\n",
    " 'mural / vascular endothelial cells 1',\n",
    " 'glutamatergic neurons 3',\n",
    " 'glutamatergic neurons 1',\n",
    " 'migrating neuroblasts',\n",
    " 'GABAergic neurons 2',\n",
    " 'glutamatergic neurons 2',\n",
    " 'radial glia 1',\n",
    " 'oligodendrocyte precursor cells',\n",
    " 'oligodendrocytes 2',\n",
    " 'microglia 2',\n",
    " 'oligodendrocytes 1',\n",
    " 'radial glia 2',\n",
    " 'glutamatergic neurons 4',\n",
    " 'oligodendrocytes 3',\n",
    " 'GABAergic neurons 3',\n",
    " 'microglia 3 / radial glia']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster red blood cells with subsample size False.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    0    0    0    0    0    2   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    1    0    0    1    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_222  E_223  E_224  E_225  E_226  E_227  E_228  E_229  E_230  \\\n",
      "0       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "1       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "2       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "3       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "...   ...  ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "4479    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4480    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4481    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4482    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4483    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "\n",
      "      E_231  \n",
      "0         0  \n",
      "1         0  \n",
      "2         0  \n",
      "3         0  \n",
      "4         0  \n",
      "...     ...  \n",
      "4479      0  \n",
      "4480      0  \n",
      "4481      0  \n",
      "4482      0  \n",
      "4483      0  \n",
      "\n",
      "[4484 rows x 360 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_227           E_227      E       227\n",
      "E_228           E_228      E       228\n",
      "E_229           E_229      E       229\n",
      "E_230           E_230      E       230\n",
      "E_231           E_231      E       231\n",
      "\n",
      "[359 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 02:21:45 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 02:21:46 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 02:21:46 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 02:33:49 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 02:33:49 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 03:04:00 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 03:35:37 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster mural / vascular endothelial cells 1 with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    0    0    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "2       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "3       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4       0  ...     0     0     0     0     0     0     0     0     1      0  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4483    0  ...     0     0     1     0     0     0     0     0     0      0  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 03:35:43 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 03:35:43 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 03:35:43 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 03:38:48 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 03:38:48 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 03:38:49 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 03:47:28 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 03:56:45 PM PDT WARNING R[write to console]: 1 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "Wed 22 Jun 2022 03:56:46 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 03:56:52 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 03:56:52 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 03:56:53 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 03:59:44 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 03:59:44 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 03:59:44 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 04:07:59 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 04:17:22 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 04:17:28 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 04:17:29 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 04:17:29 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 04:20:32 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 04:20:33 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 04:20:33 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 04:28:56 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 04:38:09 PM PDT WARNING R[write to console]: 1 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "Wed 22 Jun 2022 04:38:10 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster glutamatergic neurons 3 with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    0    0    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    2   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    1    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    2    0    0    1    0    0   \n",
      "4483          _ENSTGUG00000022543.1    1    0    0    0    0    0    1    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       0  ...     0     0     0     0     1     0     0     1     0      0  \n",
      "2       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "3       0  ...     1     0     0     0     0     0     0     0     0      0  \n",
      "4       0  ...     0     0     0     0     0     0     0     0     0      1  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4483    0  ...     1     0     0     3     0     0     0     0     1      1  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 04:38:16 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 04:38:16 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 04:38:16 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 04:41:20 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 04:41:20 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 04:50:08 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 04:59:46 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 04:59:52 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 04:59:53 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 04:59:53 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 05:02:58 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 05:02:58 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 05:02:58 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 05:12:06 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 05:22:21 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 05:22:27 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 05:22:28 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 05:22:28 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 05:25:32 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 05:25:32 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 05:25:33 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 05:34:45 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 05:44:15 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster glutamatergic neurons 1 with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    0    0    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    1    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    1    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "2       0  ...     0     0     0     2     0     0     0     0     0      0  \n",
      "3       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    3  ...     0     0     1     0     0     0     0     0     0      0  \n",
      "4483    0  ...     0     0     0     0     0     0     0     2     0      0  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 05:44:20 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 05:44:21 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 05:44:21 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 05:46:59 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 05:47:00 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 05:55:09 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 06:03:06 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 06:03:11 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 06:03:12 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 06:03:12 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 06:05:55 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 06:05:56 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 06:14:07 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 06:22:27 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 06:22:33 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 06:22:34 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 06:22:34 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 06:25:05 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 06:25:06 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 06:32:55 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 06:40:47 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster migrating neuroblasts with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    1    0    0    1    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    1    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       0  ...     0     0     0     0     0     0     1     0     0      0  \n",
      "2       0  ...     0     0     1     0     0     0     1     1     0      0  \n",
      "3       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    0  ...     1     0     0     0     0     0     0     0     0      0  \n",
      "4483    0  ...     0     0     0     1     0     0     0     0     1      0  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 06:40:52 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 06:40:53 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 06:40:53 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 06:44:08 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 06:44:09 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 06:53:49 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 07:03:58 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 07:04:04 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 07:04:04 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 07:04:04 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 07:07:18 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 07:07:18 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 07:16:24 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 07:26:10 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 07:26:16 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 07:26:17 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 07:26:17 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 07:29:20 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 07:29:20 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 07:38:17 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 07:48:02 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster GABAergic neurons 2 with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    2    0    3    0    0    0    6    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    1    0    0    0    0    0    3   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    1    0    0    0    0    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    1    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "2       0  ...     0     1     0     0     0     0     0     0     0      1  \n",
      "3       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4       0  ...     0     0     0     0     0     1     0     0     2      0  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    0  ...     0     0     0     2     0     0     0     0     0      1  \n",
      "4483    0  ...     0     0     0     0     1     0     0     0     0      0  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 07:48:08 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 07:48:09 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 07:48:09 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 07:51:21 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 07:51:21 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 08:00:18 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 08:10:49 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 08:10:55 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 08:10:55 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 08:10:55 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 08:14:01 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 08:14:01 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 08:14:01 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 08:21:42 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 08:31:41 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 08:31:46 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 08:31:47 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 08:31:47 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 08:35:01 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 08:35:01 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 08:44:16 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 08:55:04 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster glutamatergic neurons 2 with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    0    0    0    0    1    2   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    1    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       1  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "2       0  ...     0     0     0     0     0     1     0     1     0      0  \n",
      "3       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    0  ...     0     0     0     1     0     0     0     0     0      1  \n",
      "4483    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 08:55:11 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 08:55:11 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 08:55:11 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 08:58:09 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 08:58:09 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 09:06:47 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 09:15:49 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 09:15:55 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 09:15:55 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 09:15:55 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 09:18:41 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 09:18:42 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 09:26:42 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 09:34:52 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 09:34:57 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 09:34:57 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 09:34:58 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 09:37:43 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 09:37:44 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 09:45:25 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 09:53:52 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster radial glia 1 with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    2    0    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    1    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    1    0    0    1    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    2   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       0  ...     0     0     0     1     0     0     0     1     0      0  \n",
      "2       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "3       0  ...     0     1     0     0     0     0     0     0     1      0  \n",
      "4       0  ...     1     0     1     1     0     0     1     0     2      0  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4483    0  ...     0     0     0     0     0     1     0     0     0      0  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 09:53:58 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 09:53:58 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 09:53:58 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 09:57:36 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 09:57:36 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 09:57:36 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 10:07:35 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 10:19:07 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 10:19:13 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 10:19:13 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 10:19:13 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 10:22:37 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 10:22:38 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 10:22:38 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 10:31:57 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 10:43:23 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Wed 22 Jun 2022 10:43:29 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 10:43:29 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 10:43:29 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 10:46:53 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 10:46:53 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 10:46:54 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 10:55:46 PM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Wed 22 Jun 2022 11:07:23 PM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster oligodendrocyte precursor cells with subsample size False.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    1    0    0    0    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    3    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    5   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    1    0    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    1    0    0    0    1    1   \n",
      "\n",
      "      C_9  ...  E_267  E_268  E_269  E_270  E_271  E_272  E_273  E_274  E_275  \\\n",
      "0       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "1       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "2       0  ...      0      0      0      1      0      0      0      0      0   \n",
      "3       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4       0  ...      1      0      0      0      0      1      3      0      0   \n",
      "...   ...  ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "4479    9  ...      0      0      0      0      0      0      0      0      0   \n",
      "4480    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4481    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4482    0  ...      1      0      0      0      0      0      0      0      0   \n",
      "4483    1  ...      0      1      1      0      1      0      0      0      0   \n",
      "\n",
      "      E_276  \n",
      "0         0  \n",
      "1         0  \n",
      "2         0  \n",
      "3         2  \n",
      "4         1  \n",
      "...     ...  \n",
      "4479      0  \n",
      "4480      0  \n",
      "4481      0  \n",
      "4482      0  \n",
      "4483      0  \n",
      "\n",
      "[4484 rows x 475 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_272           E_272      E       272\n",
      "E_273           E_273      E       273\n",
      "E_274           E_274      E       274\n",
      "E_275           E_275      E       275\n",
      "E_276           E_276      E       276\n",
      "\n",
      "[474 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed 22 Jun 2022 11:07:35 PM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Wed 22 Jun 2022 11:07:35 PM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Wed 22 Jun 2022 11:07:36 PM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Wed 22 Jun 2022 11:35:18 PM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Wed 22 Jun 2022 11:35:18 PM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Wed 22 Jun 2022 11:35:18 PM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 12:34:39 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 01:47:53 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster oligodendrocytes 2 with subsample size False.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    0    1    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    1    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    1    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_77  E_78  E_79  E_80  E_81  E_82  E_83  E_84  E_85  E_86  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "1       0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "2       0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "3       0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "4       0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "4482    0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "4483    0  ...     0     0     0     0     0     0     0     0     0     0  \n",
      "\n",
      "[4484 rows x 214 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_82             E_82      E        82\n",
      "E_83             E_83      E        83\n",
      "E_84             E_84      E        84\n",
      "E_85             E_85      E        85\n",
      "E_86             E_86      E        86\n",
      "\n",
      "[213 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Thu 23 Jun 2022 01:47:59 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 01:48:00 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 01:48:00 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 01:51:33 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 01:51:33 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 02:03:01 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 02:13:34 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster microglia 2 with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    2    0    0    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    0    0    0    0    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    0    1    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "2       0  ...     0     0     0     0     0     0     0     1     0      0  \n",
      "3       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4       0  ...     0     0     0     1     0     0     0     0     0      0  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4483    0  ...     0     0     0     0     0     0     1     0     0      0  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Thu 23 Jun 2022 02:13:39 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 02:13:40 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 02:13:40 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 02:16:54 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 02:16:54 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 02:25:56 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 02:35:29 AM PDT WARNING R[write to console]: 2 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "Thu 23 Jun 2022 02:35:30 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Thu 23 Jun 2022 02:35:35 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 02:35:36 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 02:35:36 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 02:38:45 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 02:38:45 AM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Thu 23 Jun 2022 02:38:46 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 02:47:28 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 02:56:55 AM PDT WARNING R[write to console]: 2 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "Thu 23 Jun 2022 02:56:56 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Thu 23 Jun 2022 02:57:02 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 02:57:02 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 02:57:02 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 03:00:14 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 03:00:14 AM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Thu 23 Jun 2022 03:00:15 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 03:08:53 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 03:18:39 AM PDT WARNING R[write to console]: 3 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "Thu 23 Jun 2022 03:18:40 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster oligodendrocytes 1 with subsample size 100.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    0    0    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    1    1    0    1    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    0    1    0    0    0    0    1   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    1    0    0    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_91  E_92  E_93  E_94  E_95  E_96  E_97  E_98  E_99  E_100  \n",
      "0       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "1       0  ...     0     2     0     0     1     0     0     0     4      0  \n",
      "2       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "3       0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4       0  ...     0     0     0     0     2     0     0     2     0      1  \n",
      "...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...    ...  \n",
      "4479    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4480    0  ...     0     1     0     0     0     0     0     0     0      0  \n",
      "4481    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "4482    0  ...     0     0     0     0     0     0     1     0     0      0  \n",
      "4483    0  ...     0     0     0     0     0     0     0     0     0      0  \n",
      "\n",
      "[4484 rows x 201 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_96             E_96      E        96\n",
      "E_97             E_97      E        97\n",
      "E_98             E_98      E        98\n",
      "E_99             E_99      E        99\n",
      "E_100           E_100      E       100\n",
      "\n",
      "[200 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Thu 23 Jun 2022 03:18:46 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 03:18:46 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 03:18:46 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 03:22:04 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 03:22:04 AM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Thu 23 Jun 2022 03:22:05 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 03:30:54 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 03:40:58 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Thu 23 Jun 2022 03:41:04 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 03:41:04 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 03:41:05 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 03:44:28 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 03:44:28 AM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Thu 23 Jun 2022 03:44:29 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 03:53:41 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 04:04:13 AM PDT WARNING R[write to console]: 1 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "Thu 23 Jun 2022 04:04:14 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n",
      "Thu 23 Jun 2022 04:04:20 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 04:04:20 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 04:04:21 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 04:07:41 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 04:07:41 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 04:16:59 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 04:27:20 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster radial glia 2 with subsample size False.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    0    0    0    0    0    0    0   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    0    0    0    0    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    0    0    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    4    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    1    0    0    1    3    0    1    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    0    0    0    0    0    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    0    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    0    0    0    1    0    0    0    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_222  E_223  E_224  E_225  E_226  E_227  E_228  E_229  E_230  \\\n",
      "0       0  ...      0      1      0      0      0      0      0      0      0   \n",
      "1       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "2       0  ...      0      0      0      0      0      1      0      0      0   \n",
      "3       0  ...      0      0      0      0      0      1      0      0      0   \n",
      "4       0  ...      0      2      0      0      0      0      2      0      0   \n",
      "...   ...  ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "4479    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4480    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4481    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4482    0  ...      0      0      0      0      2      0      0      0      0   \n",
      "4483    3  ...      0      0      0      0      0      0      1      0      1   \n",
      "\n",
      "      E_231  \n",
      "0         0  \n",
      "1         0  \n",
      "2         0  \n",
      "3         0  \n",
      "4         0  \n",
      "...     ...  \n",
      "4479      0  \n",
      "4480      0  \n",
      "4481      0  \n",
      "4482      0  \n",
      "4483      0  \n",
      "\n",
      "[4484 rows x 408 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_227           E_227      E       227\n",
      "E_228           E_228      E       228\n",
      "E_229           E_229      E       229\n",
      "E_230           E_230      E       230\n",
      "E_231           E_231      E       231\n",
      "\n",
      "[407 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Thu 23 Jun 2022 04:27:30 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 04:27:31 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 04:27:31 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 04:46:14 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 04:46:14 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 05:28:33 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 06:23:17 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2 on celltype cluster glutamatergic neurons 4 with subsample size False.\n",
      "                               gene  C_1  C_2  C_3  C_4  C_5  C_6  C_7  C_8  \\\n",
      "0                           WRE_WRE    0    2    1    0    5    0    0    5   \n",
      "1       TMEM74_ENSTGUG00000022792.1    0    1    0    0    1    0    0    0   \n",
      "2     TMEM170B_ENSTGUG00000025839.1    0    2    2    0    0    0    0    0   \n",
      "3             _ENSTGUG00000020832.1    0    0    0    0    0    0    0    0   \n",
      "4             _ENSTGUG00000020221.1    0    1    0    2    1    2    0    0   \n",
      "...                             ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
      "4479          _ENSTGUG00000021313.1    0    1    0    0    0    1    0    0   \n",
      "4480     KIF5A_ENSTGUG00000025926.1    0    0    0    2    0    0    0    0   \n",
      "4481  B4GALNT1_ENSTGUG00000028363.1    0    0    0    0    0    0    0    0   \n",
      "4482     AGAP2_ENSTGUG00000019988.1    4    0    2    8    0    3    1    0   \n",
      "4483          _ENSTGUG00000022543.1    0    0    0    0    0    0    0    0   \n",
      "\n",
      "      C_9  ...  E_166  E_167  E_168  E_169  E_170  E_171  E_172  E_173  E_174  \\\n",
      "0       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "1       0  ...      2      4      0      0      0      3      0      1      0   \n",
      "2       0  ...      0      2      2      2      1      4      0      0      2   \n",
      "3       0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4       0  ...      0      0      1      0      0      0      0      0      2   \n",
      "...   ...  ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "4479    0  ...      0      0      0      0      0      0      0      0      1   \n",
      "4480    0  ...      0      0      0      0      0      0      0      0      1   \n",
      "4481    0  ...      0      0      0      0      0      0      0      0      0   \n",
      "4482    0  ...      5      7      4      6      1      8      4      0      6   \n",
      "4483    0  ...      1      0      0      0      1      0      0      0      0   \n",
      "\n",
      "      E_175  \n",
      "0         0  \n",
      "1         4  \n",
      "2         3  \n",
      "3         0  \n",
      "4         0  \n",
      "...     ...  \n",
      "4479      0  \n",
      "4480      0  \n",
      "4481      0  \n",
      "4482      2  \n",
      "4483      2  \n",
      "\n",
      "[4484 rows x 297 columns]\n",
      "           samplename sample replicate\n",
      "samplename                            \n",
      "C_1               C_1      C         1\n",
      "C_2               C_2      C         2\n",
      "C_3               C_3      C         3\n",
      "C_4               C_4      C         4\n",
      "C_5               C_5      C         5\n",
      "...               ...    ...       ...\n",
      "E_171           E_171      E       171\n",
      "E_172           E_172      E       172\n",
      "E_173           E_173      E       173\n",
      "E_174           E_174      E       174\n",
      "E_175           E_175      E       175\n",
      "\n",
      "[296 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Thu 23 Jun 2022 06:23:25 AM PDT WARNING R[write to console]: estimating size factors\n",
      "\n",
      "Thu 23 Jun 2022 06:23:26 AM PDT WARNING R[write to console]: estimating dispersions\n",
      "\n",
      "Thu 23 Jun 2022 06:23:26 AM PDT WARNING R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 06:37:18 AM PDT WARNING R[write to console]: mean-dispersion relationship\n",
      "\n",
      "Thu 23 Jun 2022 06:37:18 AM PDT WARNING R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "Thu 23 Jun 2022 06:37:18 AM PDT WARNING R[write to console]: final dispersion estimates\n",
      "\n",
      "Thu 23 Jun 2022 07:03:06 AM PDT WARNING R[write to console]: fitting model and testing\n",
      "\n",
      "Thu 23 Jun 2022 07:37:04 AM PDT INFO Using contrast: ['sample', 'C', 'E']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 17d 1h 31min 35s, sys: 16d 17h 51min 1s, total: 33d 19h 22min 37s\n",
      "Wall time: 17h 15min 29s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Define number of subsampling runs\n",
    "subsampling_runs = 3\n",
    "\n",
    "# for cluster in adata.obs.celltype.values.unique():   \n",
    "for cluster in celltypes:  \n",
    "    # Define subsample size (per condition; split 50/50 across replicates)\n",
    "    cluster_size = len(adata.obs[adata.obs[\"celltype\"]==cluster])\n",
    "    # If cluster includes > 500 cells, subsample size = 100 cells per condition\n",
    "    if cluster_size >= 500:\n",
    "        subsample_size = 100\n",
    "    # If cluster includes < 500 cells, do not subsample\n",
    "    elif cluster_size < 500 and cluster_size >= 100:\n",
    "        subsample_size = False\n",
    "    # Exclude clusters with < 100 cells\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    print(f\"Running DESeq2 on celltype cluster {cluster} with subsample size {subsample_size}.\")\n",
    "        \n",
    "    for i in np.arange(subsampling_runs):\n",
    "        ### Get control count data\n",
    "        ## Create mask to only retain cells from control 1, and one cluster\n",
    "        celltype_mask_control1 = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & [adata.obs[\"batch\"]==\"control1\"]\n",
    "        control_data1 = adata.raw[celltype_mask_control1]\n",
    "        \n",
    "        # Filter for only highly variable genes\n",
    "        control_data1 = control_data1[:,adata.var['highly_variable']]\n",
    "        \n",
    "        count_matrix_control1 = pd.DataFrame(control_data1.X.todense().T.astype(int))\n",
    "\n",
    "        if subsample_size:\n",
    "            # Randombly subsample contol 1 data\n",
    "            count_matrix_control1 = count_matrix_control1.sample(n=int(subsample_size/2), axis=\"columns\")\n",
    "            # Renumber columns \n",
    "            count_matrix_control1.columns = np.arange(int(subsample_size/2))+1\n",
    "        else:\n",
    "            # Relabel columns with indexing from 1 instead of 0\n",
    "            count_matrix_control1.columns = count_matrix_control1.columns+1\n",
    "\n",
    "        ## Create mask to only retain cells from control 2, and one cluster\n",
    "        celltype_mask_control2 = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & [adata.obs[\"batch\"]==\"control2\"]\n",
    "        control_data2 = adata.raw[celltype_mask_control2]\n",
    "        \n",
    "        # Filter for only highly variable genes\n",
    "        control_data2 = control_data2[:,adata.var['highly_variable']]\n",
    "        \n",
    "        count_matrix_control2 = pd.DataFrame(control_data2.X.todense().T.astype(int))\n",
    "\n",
    "        if subsample_size:\n",
    "            # Randombly subsample contol 2 data\n",
    "            count_matrix_control2 = count_matrix_control2.sample(n=int(subsample_size/2), axis=\"columns\")\n",
    "            # Renumber columns \n",
    "            count_matrix_control2.columns = np.arange(int(subsample_size/2))+1+(int(subsample_size/2))\n",
    "        else:\n",
    "            # Relabel columns with indexing from 1 instead of 0 + max replicate number from previous replicate\n",
    "            count_matrix_control2.columns = count_matrix_control2.columns+1+max(count_matrix_control1.columns)\n",
    "\n",
    "        # Add prefix to mark these cells as control\n",
    "        count_matrix_control1 = count_matrix_control1.add_prefix(\"C_\")\n",
    "        count_matrix_control2 = count_matrix_control2.add_prefix(\"C_\")\n",
    "        \n",
    "        ### Get experiment count data\n",
    "        ## Create mask to only retain cells from experiment 1, and one cluster\n",
    "        celltype_mask_experiment1 = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & [adata.obs[\"batch\"]==\"experiment1\"]\n",
    "        experiment_data1 = adata.raw[celltype_mask_experiment1]\n",
    "        \n",
    "        # Filter for only highly variable genes\n",
    "        experiment_data1 = experiment_data1[:,adata.var['highly_variable']]\n",
    "        \n",
    "        count_matrix_experiment1 = pd.DataFrame(experiment_data1.X.todense().T.astype(int))\n",
    "\n",
    "        if subsample_size:\n",
    "            # Randombly subsample contol 1 data\n",
    "            count_matrix_experiment1 = count_matrix_experiment1.sample(n=int(subsample_size/2), axis=\"columns\")\n",
    "            # Renumber columns \n",
    "            count_matrix_experiment1.columns = np.arange(int(subsample_size/2))+1\n",
    "        else:\n",
    "            # Relabel columns with indexing from 1 instead of 0\n",
    "            count_matrix_experiment1.columns = count_matrix_experiment1.columns+1\n",
    "\n",
    "        ## Create mask to only retain cells from experiment 2, and one cluster\n",
    "        celltype_mask_experiment2 = np.char.startswith(nd(adata.obs.celltype.values).astype(str), cluster) & [adata.obs[\"batch\"]==\"experiment2\"]\n",
    "        experiment_data2 = adata.raw[celltype_mask_experiment2]\n",
    "        \n",
    "        # Filter for only highly variable genes\n",
    "        experiment_data2 = experiment_data2[:,adata.var['highly_variable']]\n",
    "        \n",
    "        count_matrix_experiment2 = pd.DataFrame(experiment_data2.X.todense().T.astype(int))\n",
    "\n",
    "        if subsample_size:\n",
    "            # Randombly subsample contol 1 data\n",
    "            count_matrix_experiment2 = count_matrix_experiment2.sample(n=int(subsample_size/2), axis=\"columns\")\n",
    "            # Renumber columns \n",
    "            count_matrix_experiment2.columns = np.arange(int(subsample_size/2))+1+(int(subsample_size/2))\n",
    "        else:\n",
    "            # Relabel columns with indexing from 1 instead of 0 + max replicate number from previous replicate\n",
    "            count_matrix_experiment2.columns = count_matrix_experiment2.columns+1+max(count_matrix_experiment1.columns)\n",
    "\n",
    "        # Add prefix to mark these cells as experiment\n",
    "        count_matrix_experiment1 = count_matrix_experiment1.add_prefix(\"E_\")\n",
    "        count_matrix_experiment2 = count_matrix_experiment2.add_prefix(\"E_\")\n",
    "\n",
    "\n",
    "        ## Concatenate experiment and control data\n",
    "        count_matrix = pd.DataFrame()\n",
    "        count_matrix[\"gene\"] = adata.raw.var[adata.var['highly_variable']==True].index.values\n",
    "        count_matrix = pd.concat(\n",
    "            [count_matrix, count_matrix_control1, count_matrix_control2, count_matrix_experiment1, count_matrix_experiment2], axis=1\n",
    "        )\n",
    "\n",
    "        # Show the first count matrix for this cluster\n",
    "        if i==0:\n",
    "            print(count_matrix)\n",
    "\n",
    "        ### Create design matrix\n",
    "        design_matrix = pd.DataFrame()\n",
    "        design_matrix[\"samplename\"] = count_matrix.columns.values[1:]\n",
    "        # Get sample ID and replicate number from count matrix column names\n",
    "        design_matrix = design_matrix.assign(\n",
    "            sample=lambda d: d.samplename.str.extract(\"([CE])_\", expand=False)\n",
    "        )\n",
    "        design_matrix[\"replicate\"] = design_matrix[\"samplename\"].str.extractall(\"(\\d+)\").values\n",
    "        # Set sample name as index\n",
    "        design_matrix.index = design_matrix.samplename\n",
    "\n",
    "        # Show the first design matrix for this cluster\n",
    "        if i==0:\n",
    "            print(design_matrix)\n",
    "\n",
    "        ### Run DESeq2\n",
    "        dds = py_DESeq2(\n",
    "            count_matrix=count_matrix,\n",
    "            design_matrix=design_matrix,\n",
    "            design_formula=\"~ replicate + sample\",\n",
    "            gene_column=\"gene\",\n",
    "        )\n",
    "        dds.run_deseq(\n",
    "            test=\"LRT\",                            # Likelihood ratio test on the difference in deviance between a full and reduced model formula (defined by nbinomLRT)\n",
    "            reduced=\"~ replicate\",                 # Reduced model without condition as variable\n",
    "            # fitType=\"parametric\",                # Either \"parametric\", \"local\", \"mean\", or \"glmGamPoi\" for the type of fitting of dispersions to the mean intensity (default: parametric)\n",
    "            sfType=\"poscounts\",                    # Type of size factor estimation\n",
    "            betaPrior=False,                       # Whether or not to put a zero-mean normal prior on the non-intercept coefficients (default: False)\n",
    "            minReplicatesForReplace=float(\"inf\"),  # The minimum number of replicates required in order to use replaceOutliers on a sample. \n",
    "            useT=True,                             # Default is FALSE, where Wald statistics are assumed to follow a standard Normal\n",
    "            minmu=1e-06,                           # Lower bound on the estimated count for fitting gene-wise dispersion (default: 1e-06 for glmGamPoi)\n",
    "\n",
    "        )\n",
    "        dds.get_deseq_result(contrast=[\"sample\", \"C\", \"E\"])\n",
    "        res = dds.deseq_result\n",
    "        \n",
    "        res.to_csv(f\"deseq2_results_v2/deseq2_{cluster.replace(' ', '-').replace('/-', '')}_subsample_{i}.csv\")\n",
    "        \n",
    "        # Do not run again if cluster was not subsampled\n",
    "        if subsample_size is False:\n",
    "            break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each cluster, keep only the genes that are significant across all subsamples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "for cluster in adata.obs.celltype.values.unique():\n",
    "    genes_to_keep = []\n",
    "    \n",
    "    # Get all the csv files for this cluster\n",
    "    for csv in glob.glob(f\"deseq2_results_v2/deseq2_{cluster.replace(' ', '-').replace('/-', '')}_subsample_*.csv\"):\n",
    "        df = pd.read_csv(csv)\n",
    "        \n",
    "        # Bonferroni correct alpha for # subsamples x number of (highly variable) genes\n",
    "        number_of_subsamples = len(glob.glob(f\"deseq2_results_v2/deseq2_{cluster.replace(' ', '-').replace('/-', '')}_subsample_*.csv\"))\n",
    "        alpha = 0.05 / (number_of_subsamples * len(adata.var[adata.var['highly_variable']==True]))\n",
    "        \n",
    "        if len(df[df[\"padj\"] < alpha][\"padj\"]) > 0:\n",
    "            if i == 0:\n",
    "                # For the first iteration, set genes_to_keep equal to the significant genes from this subsample\n",
    "                genes_to_keep = df[df[\"padj\"] < alpha][\"gene\"].values\n",
    "                i = +1\n",
    "            else:\n",
    "                # Only keep genes that are also significant in this subsample\n",
    "                genes_to_keep = list(set(genes_to_keep) & set(df[df[\"padj\"] < alpha][\"gene\"].values))\n",
    "                \n",
    "        # If no genes are significant: empty genes to keep, break loop and move to next cluster\n",
    "        else:\n",
    "            genes_to_keep = []\n",
    "            break\n",
    "    \n",
    "    if len(genes_to_keep) > 0:\n",
    "        df_sig = pd.DataFrame()\n",
    "        df_sig[\"de_gene\"] = genes_to_keep\n",
    "        df_sig[\"cluster\"] = cluster\n",
    "        df_sig.to_csv(f\"deseq2_results_v2/significant_across_all/deseq2_{cluster.replace(' ', '-').replace('/-', '')}.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep all genes that are significant in at least one subsample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "for cluster in adata.obs.celltype.values.unique():\n",
    "    df_sig = pd.DataFrame()\n",
    "    \n",
    "    # Get all the csv files for this cluster\n",
    "    for csv in glob.glob(f\"deseq2_results_v2/deseq2_{cluster.replace(' ', '-').replace('/-', '')}_subsample_*.csv\"):\n",
    "        df = pd.read_csv(csv)\n",
    "        # Add cluster name\n",
    "        df[\"cluster\"] = cluster\n",
    "        \n",
    "        # Bonferroni correct alpha for # subsamples x number of (highly variable) genes\n",
    "        number_of_subsamples = len(glob.glob(f\"deseq2_results_v2/deseq2_{cluster.replace(' ', '-').replace('/-', '')}_subsample_*.csv\"))\n",
    "        alpha = 0.05 / (number_of_subsamples * len(adata.var[adata.var['highly_variable']==True]))\n",
    "        \n",
    "        # Add number of subsamples to df\n",
    "        df[\"number_of_subsamples\"] = number_of_subsamples\n",
    "        \n",
    "        if len(df[df[\"padj\"] < alpha][\"padj\"]) > 0:\n",
    "            if i == 0:\n",
    "                df_sig = df[df[\"padj\"] < alpha].copy()\n",
    "                i = +1\n",
    "            else:\n",
    "                df_sig = df_sig.append(df[df[\"padj\"] < alpha], ignore_index=True)\n",
    "    \n",
    "    if len(df_sig) > 0:\n",
    "        df_sig.to_csv(f\"deseq2_results_v2/significant/deseq2_{cluster.replace(' ', '-').replace('/-', '')}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
